### DISCUSSION TITLE
NEP29

### DISCUSSION THREAD

**Discussion by jamesp** [SciTools/iris#4332] [Ideas]
[NEP29 (Numpy Enhancement Proposal 29)](https://numpy.org/neps/nep-0029-deprecation_policy.html) is an approach to python package dependency management.  

In a nutshell, NEP29 states that scientific python packages should only officially support dependencies released in the last 24 months.  e.g. if you are dask and rely on numpy, the community should only expect you to support versions of numpy that have been released in the last 2 years.

Cartopy 0.20 has adopted NEP29 (see [cartopy#1856](https://github.com/SciTools/cartopy/pull/1856)).  Should we do the same in iris?

**Comment by bjlittle** [comment]
I'd personally be in favour of adopting NEP29 :+1: 

It gives explicit clarity to the community of our intentions. In addition we can be open and transparent regarding the major package dependencies that underpin and influence `iris`. 

It also gives us a clear framework to keep moving forwards, ensuring that we keep pace with the rest of the scientific python community, but it also allows us to defend minimum pins introduced to our dependencies.

I'd personally recommend the adoption of NEP29, backed up with a clear and easily discoverable (and maintained) statement of intent in our documentation.

**Comment by rcomer** [comment]
It would inform a clear policy on how long to keep workarounds (#4149).  Two years after bug is fixed in a release, we can take the workaround out.

What happens when we need to pin a dependency to an older version, e.g. `cftime`?

**Reply by bjlittle** [reply]
> What happens when we need to pin a dependency to an older version, e.g. `cftime`?

Good question. We know from experience that making a specific pin is eventually bad news. For me if we pin to a specific version, then that is a hugh red flag to the devs, and the onus is on us to unpin... or help other devs with their package (somehow) so that we can unpin.

Pragmatically, I think we review such situations given their context i.e., I'd never say never pin back, but if you do, have a plan to unpin before we hit the wall.

In this case, a live issue for `iris` is the pin `pillow<7`, which is stopping us moving to `python>3.8`... :bomb: :clock1:

**Comment by valeriupredoi** [comment]
crashing the discussion here since it looks very promising and interesting, cheers for finding this out :beer: @zklaus @bouweandela have a look at this :snake:

**Comment by wjbenfold** [comment]
Nep29 calculator tool: https://pypi.org/project/nep29/

**Comment by jamesp** [comment]
We discussed NEP29 at today's iris peloton 🚴 🚴‍♀️ 🚴‍♂️ .  An #4369 documents the questions that were raised that need to be answered for us to adopt NEP29. Please provide supporting information there, once all items are ticked off we can reevaluate.

**Comment by trexfeathers** [comment]
We have organically ended up following NEP29, so this is affectively agreed. Although it is not advertised nor automated in any way so is running on goodwill at the moment. I will close this but leave #4369 open.
